storage-dynamodb/src/main/java/zipkin2/storage/dynamodb/GetTracesForQueryCall.java

```diff
@@ -0,0 +1,261 @@
+        }
+      } else {
+        filters.add("#dur >= :min_duration");
+      }
+    }
+
+    int i = 0;
+    for (Map.Entry<String, String> tag : queryRequest.annotationQuery().entrySet()) {
+      expressionAttributeNames.put("#tag_key" + i, TAG_PREFIX + tag.getKey());
+      if (tag.getValue().isEmpty()) {
+        filterExpressionAttributes.put(":annotation" + i, new AttributeValue().withS(tag.getKey()));
+        filters.add(
+            String.format("(attribute_exists(#tag_key%d) OR contains(%s, :annotation%d))", i,
+                ANNOTATIONS, i));
+      } else {
+        filters.add(String.format("#tag_key%d = :tag_value%d", i, i));
+        filterExpressionAttributes.put(":tag_value" + i,
+            new AttributeValue().withS(tag.getValue()));
+      }
+      i++;
+    }
+
+    if (queryRequest.serviceName() != null || queryRequest.spanName() != null) {
+      // USE dynamodb query
+      if (queryRequest.serviceName() != null) {
+        if (queryRequest.spanName() != null) {
+          QueryRequest query = getQueryForGlobalSecondaryIndex(LOCAL_SERVICE_NAME_SPAN_NAME,
+              queryRequest.serviceName() + FIELD_DELIMITER + queryRequest.spanName(),
+              expressionAttributeNames, filterExpressionAttributes, filters);
+          QueryResult result = dynamoDB.query(query);
+          rows.addAll(result.getItems());
+          filterExpressionAttributes.remove(":" + LOCAL_SERVICE_NAME_SPAN_NAME);
+
+          query = getQueryForGlobalSecondaryIndex(REMOTE_SERVICE_NAME_SPAN_NAME,
+              queryRequest.serviceName() + FIELD_DELIMITER + queryRequest.spanName(),
+              expressionAttributeNames, filterExpressionAttributes, filters);
+          result = dynamoDB.query(query);
+          rows.addAll(result.getItems());
+        } else {
+          // only service names
+          QueryRequest query = getQueryForGlobalSecondaryIndex(LOCAL_SERVICE_NAME,
+              queryRequest.serviceName(), expressionAttributeNames, filterExpressionAttributes,
+              filters);
+          QueryResult result = dynamoDB.query(query);
+          rows.addAll(result.getItems());
+          filterExpressionAttributes.remove(":" + LOCAL_SERVICE_NAME);
+
+          query = getQueryForGlobalSecondaryIndex(REMOTE_SERVICE_NAME, queryRequest.serviceName(),
+              expressionAttributeNames,
+              filterExpressionAttributes, filters);
+          result = dynamoDB.query(query);
+          rows.addAll(result.getItems());
+        }
+      } else {
+        QueryRequest query = getQueryForGlobalSecondaryIndex(SPAN_NAME, queryRequest.spanName(),
+            expressionAttributeNames,
+            filterExpressionAttributes, filters);
+        QueryResult result = dynamoDB.query(query);
+        rows.addAll(result.getItems());
+      }
+    } else {
+      filters.add(TIMESTAMP_SPAN_ID + " BETWEEN :timestamp_lower AND :timestamp_upper");
+
+      ScanRequest request = new ScanRequest(spansTableName)
```

Comment:
Trying to find a way to get rid of this, Scans are bad for performance and cost. But I need some sort of key value pair (or just a key) in order to use a query.

One thought, adding another column where the timestamps are bucketed on an interval, so I have a value I can calculate from `endTs` and then work backwards one bucket at a time until I find enough trace IDs

Reply:
I don't know the technical details, but could you do something like "where trace ID is not null"? Could a fake query help

Reply:
not sure we have a solution.. punting for now